### ISSUE TITLE
ValidateSchema is not validating schema against custom keyword meta schema

### ISSUE THREAD

**Issue by nazarhussain** [ajv-validator/ajv#1221]
There is a function in Ajv API `validateSchema` which is supposed to validate any given schema against the meta schema. Hence its not celar that referred meta schema in this context is meant which is added by `addMetaSchema` or any `meta schema for custom keyword `is also validated. 

Aparently `validateSchema` its not validating against custom keyword meta schema, which ideally it should. It is validating at compile time, which is not appropriate step doing it. 

Also one related question is how to add complex validation on schema which is not possible by just defining the meta schema. e.g. In below exmaple, if I want that `range` keyword to be unique across scehma, while there can be range for any number type property. 

---

**What version of Ajv are you using? Does the issue happen if you use the latest version?**
Version: 6.12.2, yes its the latest version. 

**Ajv options object**

```javascript
const ajv = new Ajv({
    allErrors: true,
	schemaId: 'auto',
	useDefaults: false,
});

```

**Custom keyword with Meta Schema**

```javascript
ajv.addKeyword('range', {
  type: 'number',
  compile: function (sch, parentSchema) {
    var min = sch[0];
    var max = sch[1];

    return parentSchema.exclusiveRange === true
            ? function (data) { return data > min && data < max; }
            : function (data) { return data >= min && data <= max; }
  },
  errors: false,
  metaSchema: {
    type: 'array',
    items: [
      { type: 'number' },
      { type: 'number' }
    ],
    additionalItems: false
  }
});
```

**Sample Schema**

```javascript
var schema = {
  "range": [2, '4'],
  "exclusiveRange": true
};
```

**Your code**

```javascript

console.log(ajv.validateSchema(schema));
console.log(ajv.errors)

```

**Validation result, data AFTER validation, error messages**

```
Validation passes and no error. 
```

**What results did you expect?**
I was expecing error from `validateSchema` since that keyword scehma is not matched against the keyword meta schema. 

**Are you going to resolve the issue?**
Yes I can if provided the direction.

**Comment by epoberezkin** [comment]
This is by design. Adding custom keywords does not modify the meta-schema that other schemas are validated against, as custom meta-schema requires not only addition of custom keywords, but also the addition of all applicator keywords to the new meta-schema. JSON Schema 2019-09 draft proposes the two solutions to the problem of extending schemas, but both are complex and not supported in draft-07.

So validating custom keywords at the time of schema compilation is a sensible compromise. If you really need them validated at the time of schema validation, you have to create your custom meta-schema and include schemas for custom keywords there (and then you will not need to add it to the keyword definition) - the uri of this meta-schema will have to be used in $schema parameter.

**Comment by nazarhussain** [comment]
Earlier I had impression that I may be doing something wrong. But your explanation clarifies it very well. Yes, I also tried to implement all custom keyword schema as part of the parent meta schema, but it was getting complex and some edge cases were easy to handle in the custom keyword, so I choose that direction.  

Can you share a reference to proposed solution in JSON Schema 2019-09.

**Comment by epoberezkin** [comment]
> Earlier I had impression that I may be doing something wrong.

nope

> I also tried to implement all custom keyword schema as part of the parent meta schema

You may check how [hyper-schema meta-schema](https://github.com/ajv-validator/ajv/blob/master/spec/remotes/hyper-schema.json) does extension - it is not very complex if you do it manually, and has the advantage of having all your custom keywords in one place. But it doesn't work if your keywords are defined in multiple places - that is what draft 2019-09 $vocabularies idea is about... You can check how meta-schema is defined in the new draft.

> Can you share a reference to proposed solution in JSON Schema 2019-09.

`dynamicRef` is meant to simplify recursive schema extension. There is also `unevaluatedProperties` that would be needed if you would like to add keywords and also prohibit unknown keywords via meta-schema (it is managed by strictKeywords at the compilation time as well). Neither is supported atm, that's what 2019-09 issue and fund-raising is about - supporting them would require substantial re-desing and a lot of work...

**Comment by nazarhussain** [comment]
Yes, I looked into hyper-schema to get inspired, but our use case was a bit complex, and splitting the logic and meta schema into two places does not seems correct. So I kept it on place at definition of keyword and used compile to validate the schema.
